File: atcoder/codefestival/tyama_atcodercodefestival2017finalD.py
#!/usr/bin/python
n=input()
A=sorted(([int(e) for e in raw_input().split()] for _ in range(n)),key=lambda (x,y):x+y)
H=[0]
for (x,y) in A:
	for i in range(len(H)-1,-1,-1):
		if x>=H[i]:
			if i==len(H)-1: H.append(1e9)
			H[i+1]=min(H[i]+y,H[i+1])
print len(H)-1

File: atcoder/codefestival/tyama_atcodercodefestival2017finalD.rb
#!/usr/bin/ruby
gets;A=$<.map{|e|e.split.map &:to_i}.sort_by{|x,y|x+y}
H=[0]
A.map{|x,y|(s=H.size-1).downto(0){|i|x>=H[i]&&(i==s&&H<<1e9;H[i+1]>H[i]+y&&H[i+1]=H[i]+y)}}
p H.size-1
#making temporary array is slow :(
